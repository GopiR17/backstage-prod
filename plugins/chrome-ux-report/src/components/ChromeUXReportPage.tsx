/*
 * Copyright 2021 Spotify AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import {
  Page,
  Content,
  Tabs,
  useApi,
  configApiRef,
  Header
} from '@backstage/core';
import { Config } from '@backstage/config';
import { ChromeUXReportChart } from './ChromeUXReportChart';

export const ChromeUXReportPage = () => {
  const configApi = useApi(configApiRef);
  const origins = configApi.getConfigArray('chromeUXReport.origins');

  const tabs = origins.map(
    (origin: Config) => ({
      label: origin.getString("name"),
      content: <ChromeUXReportChart origin={origin.getString("site")} />,
    }),
  );

  return (
    <Page themeId="tool">
      <Header
        title="Chrome UX Report"
        subtitle="Chrome UX Report is a powerful plugin for analyzing your siteâ€™s speed in a variety of different ways."
      />
      <Content>
        <Tabs tabs={tabs} />
      </Content>
    </Page>
  );
};
