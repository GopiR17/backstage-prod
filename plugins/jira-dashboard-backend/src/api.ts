/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Config } from '@backstage/config';
import fetch from 'node-fetch';
import {
  Project,
  Filter,
  Issue,
  JiraIssue,
  NewJiraIssueResponse,
} from './types';
import { JIRA_BASE_URL_CONFIG_PATH, JIRA_TOKEN_CONFIG_PATH } from './constants';

export const getProjectInfo = async (
  projectKey: string,
  config: Config,
): Promise<Project> => {
  const response = await fetch(
    `${config.getString(JIRA_BASE_URL_CONFIG_PATH)}project/${projectKey}`,
    {
      method: 'GET',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
      },
    },
  );
  if (response.status !== 200) {
    throw Error(`${response.status}`);
  }
  return response.json();
};

export const getFilterById = async (
  id: string,
  config: Config,
): Promise<Filter> => {
  const response = await fetch(
    `${config.getString(JIRA_BASE_URL_CONFIG_PATH)}filter/${id}`,
    {
      method: 'GET',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
      },
    },
  );
  if (response.status !== 200) {
    throw Error(`${response.status}`);
  }
  const jsonResponse = await response.json();
  return { name: jsonResponse.name, query: jsonResponse.jql } as Filter;
};

export const getIssuesByFilter = async (
  projectKey: string,
  query: string,
  config: Config,
): Promise<Issue[]> => {
  const response = await fetch(
    `${config.getString(
      JIRA_BASE_URL_CONFIG_PATH,
    )}search?jql=project=${projectKey} AND ${query}`,
    {
      method: 'GET',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
      },
    },
  ).then(resp => resp.json());
  return response.issues;
};

export const getIssuesByComponent = async (
  projectKey: string,
  componentKey: string,
  config: Config,
): Promise<Issue[]> => {
  const response = await fetch(
    `${config.getString(
      JIRA_BASE_URL_CONFIG_PATH,
    )}search?jql=project=${projectKey} AND component = "${componentKey}"`,
    {
      method: 'GET',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
      },
    },
  ).then(resp => resp.json());
  return response.issues;
};

export async function getProjectAvatar(url: string, config: Config) {
  const response = await fetch(url, {
    method: 'GET',
    headers: {
      Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
    },
  });
  return response;
}

export async function postNewIssue(
  jiraIssue: JiraIssue,
  config: Config,
): Promise<NewJiraIssueResponse> {
  const body = JSON.stringify({ fields: jiraIssue });
  return await fetch(`${config.getString(JIRA_BASE_URL_CONFIG_PATH)}issue`, {
    method: 'POST',
    headers: {
      Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body,
  }).then(response => response.json());
}

export async function addIssueWatcher(
  jira_id: string,
  user: string,
  config: Config,
): Promise<void> {
  await fetch(
    `${config.getString(JIRA_BASE_URL_CONFIG_PATH)}issue/${jira_id}/watchers`,
    {
      method: 'POST',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: `"${user}"`,
    },
  );
}
