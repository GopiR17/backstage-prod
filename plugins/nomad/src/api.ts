/*
 * Copyright 2020 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Config } from '@backstage/config';
import { createApiRef } from '@backstage/core-plugin-api';

/** @public */
export const nomadApiRef = createApiRef<NomadApi>({
  id: 'plugin.nomad.service',
});

/** @public */
export type NomadApi = {
  /**
   * listAllocations is for listing all allocations matching some part of 'prefix'.
   *
   * See: https://developer.hashicorp.com/nomad/api-docs/allocations#list-allocations
   */
  listAllocations: (
    options: ListAllocationsRequest,
  ) => Promise<ListAllocationsResponse>;
};

/** @public */
export interface ListAllocationsRequest {
  namespace?: string;
  filter?: string;
}

/** @public */
export interface ListAllocationsResponse {
  allocations: Allocation[];
}

/** @public */
export interface Allocation {
  id: string;
  clientStatus: string;
  createTime: number;
  namespace: string;
  name: string;
  modifyTime: number;
  nodeID: string;
  taskGroup: string;
}

/** @public */
export interface DeploymentStatus {
  healthy: boolean;
  timestamp: string;
}

/** @public */
export class NomadHttpApi implements NomadApi {
  static fromConfig(config: Config) {
    return new NomadHttpApi(
      config.getOptionalString('nomad.nomadAddr'),
      config.getOptionalString('nomad.nomadRegion'),
      config.getOptionalString('nomad.nomadNamespace'),
      config.getOptionalString('nomad.nomadToken'),
    );
  }

  constructor(
    private nomadAddr?: string,
    private nomadRegion?: string,
    private nomadNamespace?: string,
    private nomadToken?: string,
  ) {}

  async listAllocations(
    options: ListAllocationsRequest,
  ): Promise<ListAllocationsResponse> {
    return Promise.resolve({
      allocations: [],
    });
  }
}
