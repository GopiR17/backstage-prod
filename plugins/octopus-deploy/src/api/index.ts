/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
  createApiRef,
  DiscoveryApi,
  IdentityApi,
} from '@backstage/core-plugin-api';

export type OctopusProgression = {
  Environments: OctopusEnvironment[];
  Releases: OctopusReleaseProgression[];
};

export type OctopusEnvironment = {
  Id: string;
  Name: string;
};

export type OctopusReleaseProgression = {
  Release: OctopusRelease;
  Deployments: { [key: string]: OctopusDeployment[] };
};

export type OctopusRelease = {
  Id: string;
  Version: string;
};

export type OctopusDeployment = {
  State: string;
};

export const octopusDeployApiRef = createApiRef<OctopusDeployApi>({
  id: 'plugin.octopusdeploy.service',
});

const DEFAULT_PROXY_PATH_BASE = '/octopus-deploy';

type Options = {
  discoveryApi: DiscoveryApi;
  identityApi: IdentityApi;
  /**
   * Path to use for requests via the proxy, defaults to /octopus-deploy
   */
  proxyPathBase?: string;
};

export interface OctopusDeployApi {
  getReleaseProgression(
    projectId: string,
    releaseHistoryCount: number,
  ): Promise<OctopusProgression>;
}

export class OctopusDeployClient implements OctopusDeployApi {
  private readonly discoveryApi: DiscoveryApi;
  private readonly identityApi: IdentityApi;
  private readonly proxyPathBase: string;

  constructor(options: Options) {
    this.discoveryApi = options.discoveryApi;
    this.identityApi = options.identityApi;
    this.proxyPathBase = options.proxyPathBase ?? DEFAULT_PROXY_PATH_BASE;
  }

  async getReleaseProgression(
    projectId: string,
    releaseHistoryCount: number,
  ): Promise<OctopusProgression> {
    const url = await this.getApiUrl(projectId, releaseHistoryCount);

    const { token: idToken } = await this.identityApi.getCredentials();
    const response = await fetch(url, {
      headers: idToken ? { Authorization: `Bearer ${idToken}` } : {},
    });

    let responseJson;

    try {
      responseJson = await response.json();
    } catch (e) {
      responseJson = { releases: [] };
    }

    if (response.status !== 200) {
      throw new Error(
        `Error communicating with Octopus Deploy: ${
          responseJson?.error?.title || response.statusText
        }`,
      );
    }

    return responseJson;
  }

  private async getApiUrl(projectId: string, releaseHistoryCount: number) {
    const proxyUrl = await this.discoveryApi.getBaseUrl('proxy');
    return `${proxyUrl}${this.proxyPathBase}/projects/${projectId}/progression?releaseHistoryCount=${releaseHistoryCount}`;
  }
}
