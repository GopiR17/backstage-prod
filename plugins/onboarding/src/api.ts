/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  createApiRef,
  DiscoveryApi,
  FetchApi,
  IdentityApi,
} from '@backstage/core-plugin-api';
import { CatalogApi } from '@backstage/catalog-client';

export const onboardingApiRef = createApiRef<OnboardingApi>({
  id: 'onboarding',
});

export interface OnboardingApi {
  getChecklist(groups: string, roles: string): Promise<any>;
  updateChecklistStatus(body: any): Promise<any>;
  getUserInfo(ref: string): Promise<any>;
}

export class OnboardingClient implements OnboardingApi {
  private readonly discoveryApi: DiscoveryApi;
  private readonly fetchApi: FetchApi;
  private readonly catalogApi: CatalogApi;
  constructor(options: {
    identityApi: IdentityApi;
    discoveryApi: DiscoveryApi;
    fetchApi: FetchApi;
    catalogApi: CatalogApi;
  }) {
    this.discoveryApi = options.discoveryApi;
    this.fetchApi = options.fetchApi;
    this.catalogApi = options.catalogApi;
  }

  async getUserInfo(ref: string): Promise<any> {
    return await this.catalogApi.getEntityByRef({
      kind: 'user',
      namespace: 'default',
      name: ref,
    });
  }

  async getChecklist(groups: string, roles: string): Promise<any> {
    const baseUrl = await this.discoveryApi.getBaseUrl('onboarding');
    return await this.fetchApi.fetch(
      `${baseUrl}/checklists?groups=${groups}&roles=${roles}`,
      {
        method: 'GET',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
      },
    );
  }
  async updateChecklistStatus(body: any): Promise<any> {
    const baseUrl = await this.discoveryApi.getBaseUrl('onboarding');
    return await this.fetchApi.fetch(`${baseUrl}/update-status`, {
      method: 'PUT',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }
}
