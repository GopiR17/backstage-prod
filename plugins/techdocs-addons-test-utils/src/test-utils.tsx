/*
 * Copyright 2022 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement, Fragment } from 'react';

// Shadow DOM support for the simple and complete DOM testing utilities
// https://github.com/testing-library/dom-testing-library/issues/742#issuecomment-674987855
import { screen } from 'testing-library__dom';
import { renderToStaticMarkup } from 'react-dom/server';
import { Route } from 'react-router-dom';
import { act, render } from '@testing-library/react';

import { wrapInTestApp, TestApiProvider } from '@backstage/test-utils';
import { FlatRoutes } from '@backstage/core-app-api';
import { ApiRef } from '@backstage/core-plugin-api';

import {
  TechDocsAddons,
  techdocsApiRef,
  TechDocsEntityMetadata,
  TechDocsMetadata,
  techdocsStorageApiRef,
} from '@backstage/plugin-techdocs-react';
import { TechDocsReaderPage, techdocsPlugin } from '@backstage/plugin-techdocs';
import { catalogPlugin } from '@backstage/plugin-catalog';
import { searchApiRef } from '@backstage/plugin-search-react';
import { scmIntegrationsApiRef } from '@backstage/integration-react';

const techdocsApi = {
  getTechDocsMetadata: jest.fn(),
  getEntityMetadata: jest.fn(),
};

const techdocsStorageApi = {
  getApiOrigin: jest.fn(),
  getEntityDocs: jest.fn(),
  syncEntityDocs: jest.fn(),
};

const searchApi = {
  query: jest.fn().mockResolvedValue({ results: [] }),
};

const scmIntegrationsApi = {
  fromConfig: jest.fn().mockReturnValue({}),
};

/** @ignore */
type TechDocsAddonTesterTestApiPair<TApi> = TApi extends infer TImpl
  ? readonly [ApiRef<TApi>, Partial<TImpl>]
  : never;

/** @ignore */
type TechdocsAddonTesterApis<T> = TechDocsAddonTesterTestApiPair<T>[];

type TechDocsAddonTesterOptions = {
  dom: ReactElement;
  entity: Partial<TechDocsEntityMetadata>;
  metadata: Partial<TechDocsMetadata>;
  componentId: string;
  apis: TechdocsAddonTesterApis<any>;
  path: string;
};

const defaultOptions: TechDocsAddonTesterOptions = {
  dom: <></>,
  entity: {},
  metadata: {},
  componentId: 'docs',
  apis: [],
  path: '',
};

const defaultMetadata = {
  site_name: 'Tech Docs',
  site_description: 'Tech Docs',
};

const defaultEntity = {
  kind: 'Component',
  metadata: { namespace: 'default', name: 'docs' },
};

const defaultDom = (
  <html lang="en">
    <head />
    <body>
      <div data-md-component="container">
        <div data-md-component="navigation" />
        <div data-md-component="toc" />
        <div data-md-component="main" />
      </div>
    </body>
  </html>
);

/**
 * @public
 */

export class TechDocsAddonTester {
  private options: TechDocsAddonTesterOptions = defaultOptions;
  private addons: ReactElement[];

  static buildAddonsInTechDocs(addons: ReactElement[]) {
    return new TechDocsAddonTester(addons);
  }

  private constructor(addons: ReactElement[]) {
    this.addons = addons;
  }

  withApis<T>(apis: TechdocsAddonTesterApis<T>) {
    const refs = apis.map(([ref]) => ref);
    this.options.apis = this.options.apis
      .filter(([ref]) => !refs.includes(ref))
      .concat(apis);
    return this;
  }

  withDom(dom: ReactElement) {
    this.options.dom = dom;
    return this;
  }

  withMetadata(metadata: Partial<TechDocsMetadata>) {
    this.options.metadata = metadata;
    return this;
  }

  withEntity(entity: Partial<TechDocsEntityMetadata>) {
    this.options.entity = entity;
    return this;
  }

  atPath(path: string) {
    this.options.path = path;
    return this;
  }

  build() {
    const apis: TechdocsAddonTesterApis<any> = [
      [techdocsApiRef, techdocsApi],
      [techdocsStorageApiRef, techdocsStorageApi],
      [searchApiRef, searchApi],
      [scmIntegrationsApiRef, scmIntegrationsApi],
      ...this.options.apis,
    ];

    const entityName = {
      namespace:
        this.options.entity?.metadata?.namespace ||
        defaultEntity.metadata.namespace,
      kind: this.options.entity?.kind || defaultEntity.kind,
      name: this.options.entity?.metadata?.name || defaultEntity.metadata.name,
    };

    techdocsApi.getTechDocsMetadata.mockReturnValue(
      this.options.metadata || { ...defaultMetadata },
    );
    techdocsApi.getEntityMetadata.mockResolvedValue(
      this.options.entity || { ...defaultEntity },
    );

    techdocsStorageApi.syncEntityDocs.mockResolvedValue('cached');
    techdocsStorageApi.getApiOrigin.mockResolvedValue(
      'https://backstage.example.com/api/techdocs',
    );

    techdocsStorageApi.getEntityDocs.mockResolvedValue(
      renderToStaticMarkup(this.options.dom || defaultDom),
    );

    const TechDocsAddonsPage = () => {
      return (
        <TestApiProvider apis={apis}>
          <FlatRoutes>
            <Route
              path="/docs/:namespace/:kind/:name/*"
              element={<TechDocsReaderPage />}
            >
              <TechDocsAddons>
                {this.addons.map((addon, index) => (
                  <Fragment key={index}>{addon}</Fragment>
                ))}
              </TechDocsAddons>
            </Route>
          </FlatRoutes>
        </TestApiProvider>
      );
    };

    return wrapInTestApp(<TechDocsAddonsPage />, {
      routeEntries: [
        `/docs/${entityName.namespace}/${entityName.kind}/${entityName.name}/${this.options.path}`,
      ],
      mountedRoutes: {
        '/docs': techdocsPlugin.routes.root,
        '/docs/:namespace/:kind/:name/*': techdocsPlugin.routes.docRoot,
        '/catalog/:namespace/:kind/:name': catalogPlugin.routes.catalogEntity,
      },
    });
  }

  // Components using useEffect to perform an asynchronous action (such as fetch) must be rendered within an async
  // act call to properly get the final state, even with mocked responses. This utility method makes the signature a bit
  // cleaner, since act doesn't return the result of the evaluated function.
  // https://github.com/testing-library/react-testing-library/issues/281
  // https://github.com/facebook/react/pull/14853
  async renderWithEffects(): Promise<
    typeof screen & { shadowRoot: ShadowRoot | null }
  > {
    await act(async () => {
      render(this.build());
    });

    const shadowHost = screen.getByTestId('techdocs-native-shadowroot');

    return {
      ...screen,
      shadowRoot: shadowHost?.shadowRoot || null,
    };
  }
}

export default TechDocsAddonTester.buildAddonsInTechDocs;
