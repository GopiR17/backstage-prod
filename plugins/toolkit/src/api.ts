/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
  createApiRef,
  DiscoveryApi,
  FetchApi,
  IdentityApi,
} from '@backstage/core-plugin-api';

export const toolkitApiRef = createApiRef<ToolkitApi>({
  id: 'toolkit',
});

export interface ToolkitTypo {
  id?: number;
  owner?: string;
  logo?: string;
  title?: string;
  url?: string;
  type?: string;
}
export interface ToolkitApi {
  createToolkit(body: ToolkitTypo): Promise<Response>;
  updateToolkit(body: ToolkitTypo, id: number): Promise<Response>;
  deleteOwnToolkit(id: number): Promise<Response>;
  getToolkitById(id: number): Promise<Response>;
  getMyToolkits(): Promise<Response>;
  getYourToolkits(): Promise<Response>;
  getToolkits(): Promise<Response>;
  addToolkits(body: { toolkits: number[] }): Promise<Response>;
  removeToolkit(toolkit: number): Promise<Response>;
}

export class ToolkitClient implements ToolkitApi {
  private readonly discoveryApi: DiscoveryApi;
  private readonly fetchApi: FetchApi;

  constructor(options: {
    identityApi: IdentityApi;
    discoveryApi: DiscoveryApi;
    fetchApi: FetchApi;
  }) {
    this.discoveryApi = options.discoveryApi;
    this.fetchApi = options.fetchApi;
  }

  async getYourToolkits(): Promise<Response> {
    const baseUrl = await this.discoveryApi.getBaseUrl('toolkit');
    return await this.fetchApi.fetch(`${baseUrl}/yourToolkits`);
  }

  async getMyToolkits(): Promise<Response> {
    const baseUrl = await this.discoveryApi.getBaseUrl('toolkit');
    return await this.fetchApi.fetch(`${baseUrl}/myToolkits`);
  }

  async getToolkits(): Promise<Response> {
    const baseUrl = await this.discoveryApi.getBaseUrl('toolkit');
    return await this.fetchApi.fetch(`${baseUrl}/getToolkits`);
  }

  async createToolkit(body: ToolkitTypo): Promise<Response> {
    const baseUrl = await this.discoveryApi.getBaseUrl('toolkit');
    return await this.fetchApi.fetch(`${baseUrl}/create`, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }
  async deleteOwnToolkit(id: number): Promise<Response> {
    const baseUrl = await this.discoveryApi.getBaseUrl('toolkit');
    return await this.fetchApi.fetch(`${baseUrl}/delete/${id}`, {
      method: 'DELETE',
    });
  }
  async addToolkits(body: { toolkits: number[] }): Promise<Response> {
    const baseUrl = await this.discoveryApi.getBaseUrl('toolkit');
    return await this.fetchApi.fetch(`${baseUrl}/add`, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }
  async getToolkitById(id: number): Promise<Response> {
    const baseUrl = await this.discoveryApi.getBaseUrl('toolkit');
    return await this.fetchApi.fetch(`${baseUrl}/${id}`);
  }
  async removeToolkit(toolkit: number): Promise<Response> {
    const baseUrl = await this.discoveryApi.getBaseUrl('toolkit');
    return await this.fetchApi.fetch(`${baseUrl}/remove/${toolkit}`, {
      method: 'DELETE',
    });
  }
  async updateToolkit(body: ToolkitTypo, id: number): Promise<Response> {
    const baseUrl = await this.discoveryApi.getBaseUrl('toolkit');
    return await this.fetchApi.fetch(`${baseUrl}/update/${id}`, {
      method: 'PUT',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }
}
