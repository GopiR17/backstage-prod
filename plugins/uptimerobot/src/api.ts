/*
 * Copyright 2020 Spotify AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ConfigApi, createApiRef, DiscoveryApi } from '@backstage/core';
import { ResponseError } from '@backstage/errors';
import { UptimerobotApi } from './types';

export const uptimerobotApiRef = createApiRef<UptimerobotApi>({
  id: 'plugin.uptimerobot.service',
  description: 'Used by the UptimeRobot plugin to make requests',
});

export class UptimerobotClient implements UptimerobotApi {
  public configApi: ConfigApi;
  public discoveryApi: DiscoveryApi;

  constructor({
    configApi,
    discoveryApi,
  }: {
    configApi: ConfigApi;
    discoveryApi: DiscoveryApi;
  }) {
    this.configApi = configApi;
    this.discoveryApi = discoveryApi;
  }

  private async fetch<T = any>(path: string): Promise<T> {
    const baseUrl = await this.discoveryApi.getBaseUrl('uptimerobot');

    const response = await fetch(`${baseUrl}/${path}`);
    if (!response.ok) throw await ResponseError.fromResponse(response);
    return await response.json();
  }

  getUpdateInterval(): number {
    return (
      this.configApi.getOptionalNumber('uptimerobot.updateInterval') || 120
    );
  }

  async getAllMonitors(): Promise<Response> {
    return await this.fetch<Response>('monitors');
  }

  async getSingleMonitor(annotation?: string): Promise<Response> {
    if (!annotation) return Promise.reject({ message: 'Missing annotation' });
    return await this.fetch<Response>(`monitors/${annotation}`);
  }
}
