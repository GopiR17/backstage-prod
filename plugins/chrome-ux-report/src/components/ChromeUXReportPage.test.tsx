/*
 * Copyright 2021 Spotify AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { ChromeUXReportPage } from './ChromeUXReportPage';
import { wrapInTestApp } from '@backstage/test-utils';
import { render, waitFor } from "@testing-library/react";
import {
  ApiProvider,
  ApiRegistry,
  ConfigApi,
  configApiRef,
  ConfigReader
} from "../../../../packages/core-api";
import { ChromeUXReportApi, chromeuxReportApiRef } from "../api";

describe('ChromeUXReportPage', () => {
  const configApi: ConfigApi = new ConfigReader({
    chromeUXReport: {
      origins: [{
        site: 'backstage',
        name: 'Backstage'
      }]
    },
  });

  const chromeUXReportApi: Partial<ChromeUXReportApi> = {
    getChromeUXMetrics: async () => ({
      metrics: {
        origin_id: 1,
        period_id: 1,
        connection_type: '4G',
        form_factor: 'Desktop',
        first_contentful_paint: {
          rates: { fast: 0.80, average: 0.15, slow: 0.05 }
        },
        largest_contentful_paint: {
          rates: { fast: 0.80, average: 0.15, slow: 0.05 }
        },
        dom_content_loaded: {
          rates: { fast: 0.80, average: 0.15, slow: 0.05 }
        },
        onload: {
          rates: { fast: 0.80, average: 0.15, slow: 0.05 }
        },
        first_input: {
          rates: { fast: 0.80, average: 0.15, slow: 0.05 }
        },
        layout_instability: {
          rates: { fast: 0.80, average: 0.15, slow: 0.05 }
        },
        notifications: {
          rates: { fast: 0.80, average: 0.15, slow: 0.05 }
        },
        time_to_first_byte: {
          rates: { fast: 0.80, average: 0.15, slow: 0.05 }
        },
      }
    })
  };

  const renderWrapped = (children: React.ReactNode) =>
    render(
      wrapInTestApp(
        <ApiProvider
          apis={ApiRegistry.from([
            [configApiRef, configApi],
            [chromeuxReportApiRef, chromeUXReportApi]
          ])}
        >
          {children}
        </ApiProvider>,
      ),
    );


  it('render successfully', async () => {
    const rendered = renderWrapped(<ChromeUXReportPage />);
    const headerTitle = await rendered.findByText('Chrome UX Report');
    const headerSubtitle = await rendered.findByText('Chrome UX Report is a powerful plugin for analyzing your siteâ€™s speed in a variety of different ways.');

    await waitFor(async () => {
      expect(headerTitle).toBeInTheDocument();
      expect(headerSubtitle).toBeInTheDocument();
    });
  });
});
