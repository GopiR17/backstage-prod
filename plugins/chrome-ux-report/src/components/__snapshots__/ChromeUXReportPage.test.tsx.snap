// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChromeUXReportPage render snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1"
  >
    <header
      class="makeStyles-header-2"
    >
      <div
        class="makeStyles-leftItemsBox-3"
      >
        <h4
          class="MuiTypography-root-11 makeStyles-title-5 MuiTypography-h4-19"
        >
          Chrome UX Report
        </h4>
        <h6
          class="MuiTypography-root-11 makeStyles-subtitle-6 MuiTypography-subtitle2-23"
        >
          Chrome UX Report is a powerful plugin for analyzing your siteâ€™s speed in a variety of different ways.
        </h6>
      </div>
      <div
        class="makeStyles-rightItemsBox-4"
      />
    </header>
    <article
      class="makeStyles-root-41"
    >
      <div
        class="makeStyles-root-44"
      >
        <header
          class="MuiPaper-root-58 MuiAppBar-root-47 MuiAppBar-positionStatic-51 MuiAppBar-colorPrimary-54 makeStyles-appbar-46 MuiPaper-elevation4-65"
        >
          <div>
            <div
              class="MuiTabs-root-89 makeStyles-root-88"
            >
              <div
                class="MuiTabs-scroller-94 MuiTabs-fixed-95"
                style="overflow: hidden;"
              >
                <div
                  class="MuiTabs-flexContainer-91 makeStyles-flexContainer-87"
                  role="tablist"
                >
                  <button
                    aria-selected="true"
                    class="MuiButtonBase-root-110 MuiTab-root-102 MuiTab-textColorInherit-104 makeStyles-root-100 makeStyles-root-101 Mui-selected"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiTab-wrapper-109"
                    >
                      Backstage
                    </span>
                  </button>
                </div>
                <span
                  class="PrivateTabIndicator-root-111 PrivateTabIndicator-colorSecondary-113 MuiTabs-indicator-99 makeStyles-indicator-86"
                  style="left: 0px; width: 0px;"
                >
                  <span />
                </span>
              </div>
            </div>
          </div>
        </header>
        <div
          aria-labelledby="scrollable-auto-tab-0"
          role="tabpanel"
        >
          <div
            class="MuiBox-root-115 MuiBox-root-116"
          >
            <div
              style="display: none;"
            />
          </div>
        </div>
      </div>
    </article>
  </div>
</DocumentFragment>
`;
